### PR TITLE
tests: Use fedora-with-test-tooling based on Fedora 35

### PR DESCRIPTION
**Pull request by vasiliy-ul**
**What this PR does / why we need it**:

The PR introduces fedora-with-test-tooling image based on Fedora 35 (https://github.com/kubevirt/kubevirtci/pull/713). The new version supports UEFI boot (required for https://github.com/kubevirt/kubevirt/pull/6166).

Also the new image replaces microlivecd in secure boot tests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

Major differences in Fedora 35:

- `bash` and `libreadline` in Fedora 35 come with `bracketed-paste` mode enabled by default. This results in extra escape sequences in the console output. The mode is now explicitly disabled in `/etc/profile.d/disable-bracketed-paste.sh` for all login shells (see https://github.com/kubevirt/kubevirtci/pull/713)
- Different hosts resolution modules priority in `/etc/nsswitch.conf`:
  Fedora 32: `hosts:      files dns myhostname`
  Fedora 35: `hosts:      files myhostname resolve [!UNAVAIL=return] dns`
  Here the fact that `myhostname` precedes `resolve` and `dns` modules prevents `hostname -f` from returning a `FQDN` (it just returns the short name reported by `myhostname`): https://bugzilla.redhat.com/show_bug.cgi?id=2038634

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by vasiliy-ul** [comment]
> Tests Suite: [Serial][rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system][sig-compute] VM Live Migration Starting a VirtualMachineInstance live migration cancelation should be able to cancel a migration [sig-storage][test_id:2226] with ContainerDisk

This should be fixed by https://github.com/kubevirt/kubevirt/pull/6997

**Comment by vladikr** [comment]
@EdDev @ormergi can you please take a look at the network tests?

**Comment by vladikr** [comment]
overall, look good to me, thanks @vasiliy-ul 
/approve

File: tests/vmi_configuration_test.go
Comment (EdDev) [inline tests/vmi_configuration_test.go:1379]: We are supposed to know the expect name, otherwise bugs may get in without noticing. Can this be made more explicit?

File: tests/console/login.go
Comment (EdDev) [inline tests/console/login.go:134]: Why this changed? Are the other cases even occur now (localhost and the vmi.Name)?

File: WORKSPACE
Comment (EdDev) [inline WORKSPACE:335]: We need to make sure this PR is not merged with these.

**Review by EdDev — COMMENTED** [review]
We need to wait for @oshoval and a final ack from @AlonaKaplan regarding the networking implications.
It feels like some of the findings are real bugs that fedora 35 introduced (but I am unsure).

If there is a need to push this forward faster due to non-network reasons, consider a temporary state where you use two fedora images.

Placing it on hold to emphasize that it requires more checks and to update the image registry source.

/hold

**Review by EdDev — CHANGES_REQUESTED** [review]
I think there is also a need to update https://github.com/kubevirt/kubevirtci/tree/main/cluster-provision/images/vm-image-builder/fedora-with-test-tooling with the fact we use a new image.

This is part of pushing the results to quay.
Is there such a PR on kubevirtci already?

**Comment by EdDev** [comment]
> I think there is also a need to update https://github.com/kubevirt/kubevirtci/tree/main/cluster-provision/images/vm-image-builder/fedora-with-test-tooling with the fact we use a new image.
> 
> This is part of pushing the results to quay. Is there such a PR on kubevirtci already?

Ohh found it. Missed it from the description, sorry.

File: WORKSPACE
Comment (vasiliy-ul) [reply WORKSPACE:335]: Right. We discussed that topic briefly with @rmohr in [slack](https://kubernetes.slack.com/archives/C0163DT0R8X/p1639467201175900). I pushed the image to a temporary registry in order to work on the PR and fix the issues. At the end it should be pushed to quay.io/kubevirt but I think I will probably need some help with that (from someone with proper permissions).

File: tests/console/login.go
Comment (vasiliy-ul) [reply tests/console/login.go:134]: It seems that the bash prompt has changed in Fedora 35. It prompts with `fedora` when there is no hostname.

File: tests/network/vmi_multus.go
Comment (vasiliy-ul) [reply tests/network/vmi_multus.go:1366]: In fact this should be fixed by https://github.com/kubevirt/kubevirt/pull/6997 . The NIC address info is repoted pretty quickly by guest agent. The problem is that it takes a while to propagate the update to VMI status. The delay happens because in virt-handler the doman update events are currently ignored. The mentioned PR addresses that issue. When its merged this change will not be needed. But frankly speaking I do not have good explanation why this was not observed with Fedora 32.

File: tests/vmi_configuration_test.go
Comment (vasiliy-ul) [reply tests/vmi_configuration_test.go:1379]: If fedora 35 its `Fedora Linux`. I thought it would be enough to search only for the `Fedora` substring. That will make the test less 'fragile' and alow easier transition to new images.

File: tests/network/vmi_subdomain.go
Comment (vasiliy-ul) [reply tests/network/vmi_subdomain.go:159]: Yep, /etc/nsswitch.conf in Fedora 35 looks a bit wierd. Just if you havent checked already: I put my findings on this in the PR description.

Comment (oshoval) [reply tests/network/vmi_subdomain.go:159]: Lets do this sed as part of the provision [0] ?
I wonder if it will affect the behavior of [1] as well

Note that for all the machines that have `CloudInitNoCloud` we do provide a hostname in the metadata of cloud-init (done internally by kubevirt [2])
so better to make sure that this feature is not broken by fedora 35 and masked by the change [1]
(however not all the tests use a fedora with cloud-init)
Changing as suggested on [0] would make the flow the same for all the tests, and more predictable imo

[0] https://github.com/kubevirt/kubevirtci/pull/713/files#diff-60e708f2e40cf0f5d49420a079fdd43d2a2b2691cada7e745ee1033d195d8a80R44
[1] https://github.com/kubevirt/kubevirt/pull/6958#discussion_r779701844
[2] https://github.com/kubevirt/kubevirt/blob/7c10ae70135b9262122feab31db2026d756f76b5/pkg/cloud-init/cloud-init.go#L361

**Comment by oshoval** [comment]
> If there is a need to push this forward faster due to non-network reasons, consider a temporary state where you use two fedora images.

+1, no need to rush consolidation so fast imho

File: tests/network/vmi_multus.go
Comment (EdDev) [reply tests/network/vmi_multus.go:1366]: I understood that the ref PR is not closed, but generally speaking, whenever a timer is increased it feels like we are masking a real problem. We have many other tests that are waiting for the guest-agent to report back to the VMI status, so changing it only here is too odd.

I would feel much better if this will not be needed under this change, but fix it in a pre-change. Said that, I also do not feel comftable in blocking this on an unrelated potential bug.

**Comment by vasiliy-ul** [comment]
> > If there is a need to push this forward faster due to non-network reasons, consider a temporary state where you use two fedora images.
> 
> +1, no need to rush consolidation so fast imho

Agree, no need to rush with that if it may result in additional CI unstability. I am in principle fine with introducing a new image and thus having a 'transition period'. The reason for that whole activity is UEFI boot in Fedora 35. It will allow replacing microlivecd in SecureBoot tests. Also UEFI is required for SEV testing (which at the moment lacks the hardware anyway). @vladikr , @rmohr , @jean-edouard any thoughts on that?

File: tests/network/vmi_subdomain.go
Comment (EdDev) [reply tests/network/vmi_subdomain.go:159]: > Yep, /etc/nsswitch.conf in Fedora 35 looks a bit wierd. Just if you havent checked already: I put my findings on this in the PR description.

Well, we are suppose to take new distributions in early to confirm we can integrate well with them as guests. However, if the problems comes from the distribution, I am not sure if it is good for "us" to adjust to something wrong.
This specific tests is interested in the correct hostname to be use, so manipulating the guest feels wrong to me.
I guess, that is a good enough reason to leave this test working with the previous image.

However, maybe we can confirm this is indeed a bug by opening a BZ against Fedora and if this is confirmed we could patch it in the provision stage with a ref to the bug and an issue that tracks it (so we can revert the patch).

Not sure what is the next best action to be taken here (two images, or ref a BZ and patching the distribution).

File: tests/vmi_configuration_test.go
Comment (EdDev) [reply tests/vmi_configuration_test.go:1379]: I understand, but on the other side it may masks bugs because it is too loose.
It is worth another opinion on this from the ones who maintain these tests.

File: tests/network/vmi_multus.go
Comment (vasiliy-ul) [reply tests/network/vmi_multus.go:1366]: > whenever a timer is increased it feels like we are masking a real problem

I fully agree here. I got some feedback on that PR and seems that the proposed solution was not the right one. I could clearly observe that the guest got the address for the secondary NIC rather quickly and then GA reported it. But the update of the vmi status was taking too long. I will need to debug it further.

File: tests/network/vmi_subdomain.go
Comment (vasiliy-ul) [reply tests/network/vmi_subdomain.go:159]: Diving a bit more into the topic... found the following: `man nss-resolve` [1] states:

> it is recommended to place "resolve" early in /etc/nsswitch.conf's "hosts:" line. It should be before the "files" entry, since systemd-resolved supports /etc/hosts internally, but with caching. To the contrary, it should be after "mymachines", to give hostnames given to local VMs and containers precedence over names received over DNS. Finally, we recommend placing "dns" somewhere after "resolve", to fall back to nss-dns if systemd-resolved.service is not available.

then `man nss-myhostname` [2]:

> It is recommended to place "myhostname" after "file" and before "dns". This resolves well-known hostnames like "localhost" and the machine hostnames locally. It is consistent with the behaviour of nss-resolve, and still allows overriding via /etc/hosts.

So the recommended **correct** `hosts:` line in /etc/nsswitch.conf should look like the following:

> mymachines resolve [!UNAVAIL=return] **files myhostname** dns

While in Fedora 35 it is:

> **files myhostname** resolve [!UNAVAIL=return] dns

So it seems that `resolve` is positioned wrong here. With `resolve [!UNAVAIL=return] **files myhostname** dns` fqdn works correctly.

[1] https://www.freedesktop.org/software/systemd/man/nss-resolve.html
[2] https://www.freedesktop.org/software/systemd/man/nss-myhostname.html

**Comment by vladikr** [comment]
> > > If there is a need to push this forward faster due to non-network reasons, consider a temporary state where you use two fedora images.
> > 
> > 
> > +1, no need to rush consolidation so fast imho
> 
> Agree, no need to rush with that if it may result in additional CI unstability. I am in principle fine with introducing a new image and thus having a 'transition period'. The reason for that whole activity is UEFI boot in Fedora 35. It will allow replacing microlivecd in SecureBoot tests. Also UEFI is required for SEV testing (which at the moment lacks the hardware anyway). @vladikr , @rmohr , @jean-edouard any thoughts on that?

@vasiliy-ul in general, I'm not against a separate image if this unblocks your future work. What is the plan for ending this `transition period`? ;)
What are the outstanding issues?
Do I read correctly that the issue with the `nsswitch` is resolvable?

**Comment by vasiliy-ul** [comment]
> What is the plan for ending this `transition period`? ;) What are the outstanding issues?

That is a good question. The tests that are currently enabled pass as can be seen in this PR. I think there are concerns about more potential issues from the network perspective (considering the nsswitch problem).  Also there many `XFAIL` tests. Not sure what to do with those TBH. Since they are known to fail anyway its hard to tell if the new image will affect them somehow.

> Do I read correctly that the issue with the `nsswitch` is resolvable?

Yes, it is resolvable by adjusting `/etc/nsswitch.conf` (currently it is done within this PR but can be moved to the image provision step).

File: tests/network/vmi_subdomain.go
Comment (oshoval) [reply tests/network/vmi_subdomain.go:159]: Nice, i wonder just what /etc/hosts has, 
maybe it has a value that didn't exists on previous fedora, and it cause the order mess ?

Comment (vasiliy-ul) [reply tests/network/vmi_subdomain.go:159]: > i just wonder what /etc/hosts has

hm, nothing special IMHO:

```bash
[fedora@fc35 ~]$ cat /etc/hosts
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6
[fedora@fc35 ~]$ grep "hosts: " /etc/nsswitch.conf
hosts:      files myhostname resolve [!UNAVAIL=return] dns
[fedora@fc35 ~]$ hostname -f
fc35
[fedora@fc35 ~]$ sudo sed -i -e 's/myhostname//g' /etc/nsswitch.conf && hostname -f
fc35.testsubdomain.default.svc.cluster.local
```

Comment (oshoval) [reply tests/network/vmi_subdomain.go:159]: Thanks for checking

Comment (vasiliy-ul) [reply tests/network/vmi_subdomain.go:159]: > Thanks for checking

:ok_hand: 

... and now there is an open bug for systemd [1] along with proposed fix [2]

[1] https://github.com/systemd/systemd/issues/22081
[2] https://github.com/systemd/systemd/pull/22078

File: tests/network/vmi_multus.go
Comment (vasiliy-ul) [reply tests/network/vmi_multus.go]: This is fixed by https://github.com/kubevirt/kubevirt/pull/7131

File: tests/network/vmi_subdomain.go
Comment (vasiliy-ul) [reply tests/network/vmi_subdomain.go]: Hey @EdDev , @oshoval , @AlonaKaplan , what do you think about this? I moved the patching of `/etc/nsswitch.conf` to the provision step [1] so no need to adjust the tests here anymore. There is already a BZ against fedora [2]. Other than that I do not observe any problems with the new image. After several rebases the test runs seem quite stable. Would it be fine to go forward with this PR or are you still unsure and have concerns about it?

[1] https://github.com/kubevirt/kubevirtci/pull/713
[2] https://bugzilla.redhat.com/show_bug.cgi?id=2038634

Comment (oshoval) [reply tests/network/vmi_subdomain.go]: I am fine with it, thanks, lets see other opinions if needed,
maybe worth to add a comment in the test as well  (referencing the change on kubevirtci, that has all the info) ?
so once it is fixed by fedora, in case it will breaks it will be visible easily in the test?
(not sure it will break the test, but if so)

Comment (jean-edouard) [reply tests/network/vmi_subdomain.go]: LGTM
@vasiliy-ul will the upstream fix break any test? In obvious ways?
@EdDev @AlonaKaplan any additional concern?

Comment (vasiliy-ul) [reply tests/network/vmi_subdomain.go]: >  @vasiliy-ul will the upstream fix break any test? In obvious ways?

If the fix for systemd is accepted it will affect only `nss-myhostame`. Cannot thinnk of any issue with that. Should work even with the patched `nsswitch.conf`. We just change the priority of the resolve modules there.

**Comment by EdDev** [comment]
I am a bit lost here, so I am sorry if I the following concerns have already been treated.

Are we sure this patching done to the base fedora will not mask an issue that a future user may face?
That is my main concern here. If Fedora will keep this, some poor user using it may have a problem when running a guest under kubevirt.

**Comment by vasiliy-ul** [comment]
> Are we sure this patching done to the base fedora will not mask an issue that a future user may face?
That is my main concern here. If Fedora will keep this, some poor user using it may have a problem when running a guest under kubevirt.

I believe its a valid concern. For sure there are plenty of issues inside the guests that may break functionality of KubeVirt. Though anyway its not realistic to test all the possible OSes here in CI. The patching of the base fedora image is just one of the many possible configurations. If we decide that its valid then I do not see problems here. IMHO *any* OS configuration may bring issues... now or later at some point.

**Comment by rmohr** [comment]
> > What is the plan for ending this `transition period`? ;) What are the outstanding issues?
> 
> That is a good question. The tests that are currently enabled pass as can be seen in this PR. I think there are concerns about more potential issues from the network perspective (considering the nsswitch problem). Also there are many `XFAIL` tests. Not sure what to do with those TBH. Since they are known to fail anyway its hard to tell if the new image will affect them somehow.
> 

I think we should delete the XFAIL tests and remove XFail.

> > Do I read correctly that the issue with the `nsswitch` is resolvable?
> 
> Yes, it is resolvable by adjusting `/etc/nsswitch.conf` (currently it is done within this PR but can be moved to the image provision step).


I think that independent of fc35 having issues or not, guests work with certain configurations better with kubevirt than with others. If we can adjust kubevirt to work better with a broader set of default images better, let's definitely track it and improve kubevirt. If we detect a bug in a guest ,we can definitely work around it and file an issue on their side. That some guest may have issues should be irrelevant for the tested functionality.

So, from my point of view, as soon as we have ensured that we track any issues/RFEs encountered, I would be happy to make progress with this PR.

**Comment by vasiliy-ul** [comment]
> So, from my point of view, as soon as we have ensured that we track any issues/RFEs encountered, I would be happy to make progress with this PR.

There are already bugzilla [1] and github [2] issues open. Also there is a proposed fix for systemd [3].

[1] https://bugzilla.redhat.com/show_bug.cgi?id=2038634
[2] https://github.com/systemd/systemd/issues/22081
[3] https://github.com/systemd/systemd/pull/22078

**Comment by rmohr** [comment]
> > So, from my point of view, as soon as we have ensured that we track any issues/RFEs encountered, I would be happy to make progress with this PR.
> 
> There are already bugzilla [1] and github [2] issues open. Also there is a proposed fix for systemd [3].
> 
> [1] https://bugzilla.redhat.com/show_bug.cgi?id=2038634 [2] [systemd/systemd#22081](https://github.com/systemd/systemd/issues/22081) [3] [systemd/systemd#22078](https://github.com/systemd/systemd/pull/22078)

@vasiliy-ul @AlonaKaplan, so do you think that we understand all issues we faced and we can remove the hold? It looks so to me.

**Comment by vasiliy-ul** [comment]
I added the details about the major issues in the PR description. All they are fixed in the base image now https://github.com/kubevirt/kubevirtci/pull/713 . So I believe the hold can be removed. The only thing left here is the refernce to my testing repo `repository = "vasiliy-ul/fedora-with-test-tooling",`.

**Comment by vasiliy-ul** [comment]
Hey, ping here. Just wanted to bring this up: are there any more concerns preventing from moving this forward?

**Comment by rmohr** [comment]
/unhold
/lgtm

let's see if tests need a rebase, but looks good otherwise.

**Comment by jean-edouard** [comment]
@rmohr thank you for pushing an updated image to kubevirtci! Shouldn't do the aarch64 one too?

**Comment by rmohr** [comment]
> @rmohr thank you for pushing an updated image to kubevirtci! Shouldn't we do the aarch64 one too?

Yes. But tests are passing for now on arm64. We can bring in the optimizations later. @zhlhahaha FYI. I want to finally remove the special image need for EFI. What do you think?

**Comment by zhlhahaha** [comment]
> > @rmohr thank you for pushing an updated image to kubevirtci! Shouldn't we do the aarch64 one too?
> 
> Yes. But tests are passing for now on arm64. We can bring in the optimizations later. @zhlhahaha FYI. I want to finally remove the special image need for EFI. What do you think?

Do you means we should have same base image for both arm64 and x86_64 for testing? If my understanding is right, I used to send a PR for build multi-arch base image. https://github.com/kubevirt/kubevirtci/pull/732

**Comment by vasiliy-ul** [comment]
All storage lanes were failing on the same tests. Rebased the PR to run it with the latest code.

**Comment by brybacki** [comment]
It looks like the first few times storage tests failed on Volume size, so what is the actual size of the image?
I can see that now tests fail on some assertions. Maybe old assumptions need to be updated. I can look into fstrim test.

**Comment by vasiliy-ul** [comment]
> It looks like the first few times storage tests failed on Volume size, so what is the actual size of the image? I can see that now tests fail on some assertions. Maybe old assumptions need to be updated. I can look into fstrim test.

Right, the first time the test were requesting `5Gi` volume size while with the new Fedora it should be `6Gi`. That I fixed in the latest patch.

Now seems that this part needs adjustments as well:

https://github.com/kubevirt/kubevirt/blob/9da70587e2e76ced7eebab1877fd627c257b9a80/tests/storage/datavolume.go#L965-L973

Since the volume size has increased, probably there is a need to fill more space

**Comment by brybacki** [comment]
> adjustments

I see a different failure:

```
Expected
    <int64>: 1276432
to be >
    <int64>: 1276432
tests/storage/datavolume.go:979
```

**Comment by brybacki** [comment]
@maya-r what is the idea here? can you help?

**Comment by vasiliy-ul** [comment]
> > adjustments
> 
> I see a different failure:
> 
> ```
> Expected
>     <int64>: 1276432
> to be >
>     <int64>: 1276432
> tests/storage/datavolume.go:979
> ```

Its without preallocation I think.

**Comment by brybacki** [comment]
> 



> 

The message is strange value looks good. Maybe the timeout? But why so slow?

**Comment by brybacki** [comment]
> Since the volume size has increased, probably there is a need to fill more space

You think there is some space reserved? Or maybe trim was not used why creating the image? In that case filling bigger image would help. But honestly, I am not sure how it works.

**Comment by vasiliy-ul** [comment]
I tuned the test parameters and locally everything passes. Though honestly I do not know if there is some 'clever' way of picking the proper values. I just did several experimental runs and.

**Comment by brybacki** [comment]
> I tuned the test parameters and locally everything passes. Though honestly I do not know if there is some 'clever' way of picking the proper values. I just did several experimental runs.

Anyway - It would be good to have fedora image with smaller footprint.

**Comment by vasiliy-ul** [comment]
Seems that CI is completely stuck here. How to retrigger it?

**Comment by oshoval** [comment]
> Seems that CI is completely stuck here. ~How to retrigger it?~ ...ah, nm, just force-pushed...

Maybe its not just this PR, rebase or commit amend without changes will do

**Comment by vasiliy-ul** [comment]
Yeah, did --amend --no-edit. But still I see that some jobs are not triggered. E.g.

`pull-kubevirt-e2e-kind-1.22-sriov Pending — `
`pull-kubevirt-e2e-kind-1.22-sriov-nonroot Pending —  `
`pull-kubevirt-e2e-kind-1.23-vgpu Pending —  `
`pull-kubevirt-e2e-kind-1.23-vgpu-nonroot Pending —   `
`pull-kubevirt-e2e-windows2016 Pending —    `

Is it some outage in CI?

**Comment by oshoval** [comment]
> Is it some outage in CI?

Seems so yes, it is not just this PR

**Review by jean-edouard — APPROVED** [review]
/test pull-kubevirt-e2e-kind-1.23-vgpu-nonroot
/test pull-kubevirt-e2e-k8s-1.23-sig-network-nonroot

**Pull request merged**
